fem/estimators.cpp

```diff
@@ -207,123 +211,106 @@ void KellyErrorEstimator::ComputeEstimates()
-         mesh->GetFaceInfos(f, &Inf1, &Inf2, &NCFace);
-
-         // Convention
-         // * Conforming face: Face side with smaller element id handles
-         // the integration
-         // * Non-conforming face: The slave handles the integration.
-         // See FaceInfo documentation for details.
-         bool isNCSlave    = FT->Elem2No >= 0 && NCFace >= 0;
-         bool isConforming = FT->Elem2No >= 0 && NCFace == -1;
-         if ((FT->Elem1No < FT->Elem2No && isConforming) || isNCSlave)
+         // Compute NC and face information
+         int FaceElement1, FaceElement2, NCFace;
+         mesh->GetFaceInfos(fi, &FaceElement1, &FaceElement2, &NCFace);
+         mesh->GetFaceElements(fi, &FaceElement1, &FaceElement2);
+
+         // We skip over master faces
+         bool isNCSlave    = FaceElement2 >= 0 && NCFace >= 0;
+         bool isConforming = FaceElement2 >= 0 && NCFace == -1;
+         if (isConforming || isNCSlave)
          {
             if (attributes.Size() &&
-                (attributes.FindSorted(FT->Elem1->Attribute) == -1
-                 || attributes.FindSorted(FT->Elem2->Attribute) == -1))
+                (attributes.FindSorted(mesh->GetAttribute(FaceElement1)) == -1
+                 || attributes.FindSorted(mesh->GetAttribute(FaceElement2)) == -1))
             {
                continue;
             }
 
-            IntegrationRule eir;
-            Vector jumps(nip);
+            // Computation of the transformations is not super cheap, so we delay it until here. All ops above are cheaper lookups.
+            auto FT = mesh->GetFaceElementTransformations(fi);
+
+            // TODO Refactor this computation into a user-facing function.
+            auto &int_rule = IntRules.Get(FT->FaceGeom, 2 * xfes->GetFaceOrder(fi));
+            const auto nip = int_rule.GetNPoints();
+
+            double jump_integral = 0.0;
 
             // Integral over local half face on the side of e₁
             // i.e. the numerical integration of ∫ flux ⋅ n dS₁
             for (int i = 0; i < nip; i++)
             {
-               // Evaluate flux at IP
+               // Set up integration point
                auto &fip = int_rule.IntPoint(i);
-               IntegrationPoint ip;
-               FT->Loc1.Transform(fip, ip);
-
-               Vector val(flux_space->GetVDim());
-               flux->GetVectorValue(FT->Elem1No, ip, val);
-
-               // And build scalar product with normal
-               Vector normal(mesh->SpaceDimension());
                FT->Face->SetIntPoint(&fip);
+
+               // Compute normal - note that the normals match at each face integration point up to the sign!
                if (mesh->Dimension() == mesh->SpaceDimension())
                {
+                  // This computes a weighted normal, so we divide by the weight to get back the normal.
                   CalcOrtho(FT->Face->Jacobian(), normal);
+                  normal /= FT->Face->Weight();
```

Comment:
I agree that the normal of the face should be divided by the Weight, to get a unit normal vector.